keras_hub/src/models/parseq/parseq_tokenizer.py

```diff
@@ -0,0 +1,142 @@
+
+        self.remove_whitespace = remove_whitespace
+        self.normalize_unicode = normalize_unicode
+        self.max_label_length = max_label_length
+        self._add_special_token("[B]", "start_token")
+        self._add_special_token("[E]", "end_token")
+        self._add_special_token("[P]", "pad_token")
+        # Create lookup tables.
+        self.char_to_id = tf.lookup.StaticHashTable(
+            initializer=tf.lookup.KeyValueTensorInitializer(
+                keys=list(self._stoi.keys()),
+                values=list(self._stoi.values()),
+                key_dtype=tf.string,
+                value_dtype=tf.int32,
+            ),
+            default_value=0,
+        )
+        self.id_to_char = tf.lookup.StaticHashTable(
+            initializer=tf.lookup.KeyValueTensorInitializer(
+                keys=list(self._stoi.values()),
+                values=list(self._stoi.keys()),
+                key_dtype=tf.int32,
+                value_dtype=tf.string,
+            ),
+            default_value=self.pad_token,
+        )
+
+    def id_to_token(self, id):
+        if id >= self.vocabulary_size() or id < 0:
+            raise ValueError(
+                f"`id` must be in range [0, {self.vocabulary_size() - 1}]. "
+                f"Received: {id}"
+            )
+        return self._itos[id]
+
+    def token_to_id(self, token):
+        return self._stoi[token]
+
+    def _preprocess(self, label):
+        """Performs preprocessing include only characters from ASCII."""
+        if self.remove_whitespace:
+            label = tf.strings.regex_replace(label, r"\s+", "")
+
+        if self.normalize_unicode:
+            label = tf_text.normalize_utf8(label, normalization_form="NFKD")
+            label = tf.strings.regex_replace(label, r"[^!-~]", "")
+
+        if self.lowercase_only:
+            label = tf.strings.lower(label)
+        elif self.uppercase_only:
+            label = tf.strings.upper(label)
+
+        label = tf.strings.regex_replace(label, self.unsupported_regex, "")
+        label = tf.strings.substr(label, 0, self.max_label_length)
+
+        return label
+
+    @preprocessing_function
+    def tokenize(self, inputs):
+        self._check_vocabulary()
+        inputs = tf.convert_to_tensor(inputs)
+        unbatched = inputs.shape.rank == 0
+        if unbatched:
+            inputs = tf.expand_dims(inputs, 0)
+
+        inputs = tf.map_fn(self._preprocess, inputs, dtype=tf.string)
+
+        if tf.size(inputs) > 0:
+            chars = tf.strings.unicode_split(inputs, "UTF-8")
+            token_ids = self.char_to_id.lookup(chars)
+            token_ids = tf.cast(token_ids, dtype=tf.int32)
+        else:
+            token_ids = tf.ragged.constant([], dtype=tf.int32)
+
+        return token_ids
+
+    def vocabulary_size(self):
+        """Get the integer size of the tokenizer vocabulary."""
+        self._check_vocabulary()
+        return len(self.vocabulary)
```

Comment:
Shouldn't this be `len(self.vocabulary) + 3` because you are adding 3 special tokens?

Reply:
Yeah make sense actually.